### PR TITLE
Added split operations to IsSequence, EqSequence

### PR DESCRIPTION
**Pull request by w3rs**

### CODE REVIEW

File: src/Data/Sequences.hs [hunk 1]
```diff
@@ -411,6 +410,12 @@ class (Monoid seq, MonoTraversable seq, SemiSequence seq, MonoPointed seq) => Is
     unsafeIndex :: seq -> Index seq -> Element seq
     unsafeIndex = indexEx
 
+
+    -- | 'splitWhen' splits a sequence into components delimited by separators, where the
+    -- predicate returns True for a separator element.
```
Comment (snoyberg) [inline src/Data/Sequences.hs]: Would you mind adding a comment about whether the separators are retained or discarded? A concrete example would be useful. At least I personally get confused about this whenever I use split-style functions.

File: src/Data/Sequences.hs [hunk 2]
```diff
@@ -1198,6 +1219,17 @@ instance VS.Storable a => IsSequence (VS.Vector a) where
 
 -- | A typeclass for sequences whose elements have the 'Eq' typeclass
 class (MonoFoldableEq seq, IsSequence seq, Eq (Element seq)) => EqSequence seq where
+
+    -- | @'split' sep@ splits a sequence into components delimited by separator
+    -- element.
```
Comment (snoyberg) [inline src/Data/Sequences.hs]: Same documentation comment from above applies here.

**Comment by snoyberg** [comment]
In addition to the documentation comments, would you be able to add a test case for this, to codify the semantics?

**Comment by w3rs** [comment]
I'll add following (from ByteString docs):

> The resulting components do not contain the separators. Two adjacent separators result in an empty component in the output.

And will work on test case as well

**Comment by gregwebs** [comment]
I find the naming of these unintuitive (even if it may be consistent with some of the underlying libraries).
Our current naming scheme uses `on` to signify a function for `groupOn` (in that case it is returning an Ord). So `splitOn` and `splitWhen` `when` makes sense, bu

**Comment by gregwebs** [comment]
I find the naming of these unintuitive (even if it may be consistent with some of the underlying libraries).
Our current naming scheme uses `on` to signify a function for `groupOn` (in that case it is returning an Ord). So `splitOn` and `splitWhen` would mean the same thing by that convention

I think `split` from the patch can be renamed to `splitBy` or `splitAt` or `splitElem`.
Perhaps `splitOn` from the patch can be renamed to `splitSeq` ?

**Comment by MaxGabriel** [comment]
IIRC the different splitting operators have slightly different edge case behavior (IIRC it was how they handle being given an empty string as the separator, or how splitting on a string consisting of just the separator works). I don't remember for sure but it might be worth double checking.

**Comment by w3rs** [comment]
Considering the naming, `Data.List.Split`, `ByteString` and `Text` provide three different naming schemes. I followed the first, except I added lacking `split :: Element seq -> seq -> [seq]`. I find myself using `split` most of time, especially for CSV/TSV parsing, so I judged it deserves the shortest name.

I think the most intuitive short name for `seq -> seq -> [seq]` must be `splitSeq`. Following the same logic, I could rename `split` to `splitElem`, but then there is no short `split` name for anything.

To summarize:

`splitWhen` is good name already
`splitOn` should be renamed to `splitSeq` (better ideas?)
`split` should either stay same or be renamed to `splitElem`

**Comment by gregwebs** [comment]
@w3rs I agree with your explanation. Do you use classy-prelude? One option is to re-export `splitElem` as `split` in classy-prelude. I am still wondering if there is something more intuitive: `splitAt` captures it for me, except that `at` often refers to integer indexed.

**Comment by w3rs** [comment]
> Do you use classy-prelude? One option is to re-export splitElem as split in classy-prelude.

@gregwebs yes, it makes sense.

> splitAt captures it for me, except that at often refers to integer indexed.

This name is actually already taken with this purpose:

    splitAt :: Index seq -> seq -> (seq, seq)

**Comment by gregwebs** [comment]
I am favoring `splitElem`, but I will go with consensus (or better ideas that come up).

**Comment by w3rs** [comment]
I pushed the new commit with improvements. It also has new `intercalate` function for `IsSequence` (I needed it as inverse for splitSeq)

**Comment by gregwebs** [comment]
Looking good! If you want a shorter name, you could use `splitEq` instead of `splitElem`

**Comment by w3rs** [comment]
@gregwebs I think I can bear a couple extra chars after all.

**Comment by snoyberg** [comment]
Looks good to me, I'm happy to merge thanks for all the improvements.

Can you both confirm that this is ready to go?

**Comment by w3rs** [comment]
@snoyberg since we couldn't find much better names than `splitElem` and `splitSeq`, I'm fine with what we have.

File: src/Data/Sequences.hs [hunk 3]
```diff
@@ -1198,6 +1244,38 @@ instance VS.Storable a => IsSequence (VS.Vector a) where
...
+    splitElem x = splitWhen (== x)
+
+    -- | @'splitSeq'@ splits a sequence into components delimited by
+    -- separator subsequence. 'splitSeq' is the right inverse of 'intercalate':
+    --
+    -- > intercalate x . splitSeq x === id
+    --
+    -- 'splitElem' can be considered a special case of 'splitSeq'
+    --
+    -- > 'splitSeq' (singleton sep) === 'splitElem sep'
```
Comment (kbillings) [inline src/Data/Sequences.hs]: You cannot use `''` with `>`. It produces invalid documentation. You must wrap the statement like:
```
@
'splitSeq' (singleton sep) === 'splitElem' sep
@
```

**Comment by w3rs** [comment]
@kbillings good catch! I had no success with `stack haddock` forgot to ask others to review the docs. Will replace `>` with `@`

**Comment by gregwebs** [comment]
thanks a lot!
